File: examples/companion-packages/dotnet-csharp-app/src/AgentXM.Examples.PawMatch.CSharp/Variants.cs
namespace AgentXM.Examples.PawMatch.CSharp;

internal enum PetCardStyle
{
    Compact,
    Detailed,
    Playful,
}

internal enum MatchStrategy
{
    Popularity,
    MatchQuiz,
    LongestStay,
}

internal enum MatchDepth
{
    Short,
    Standard,
    Thorough,
}

internal enum DonateFocus
{
    All,
    Shelters,
    Rescue,
}

internal static class Variants
{
    public static T Parse<T>(string value) where T : struct, Enum
    {
        Span<char> buffer = stackalloc char[value.Length];
        var length = 0;
        var capitalize = true;
        foreach (var ch in value)
        {
            if (ch == '-')
            {
                capitalize = true;
                continue;
            }
            buffer[length++] = capitalize ? char.ToUpperInvariant(ch) : ch;
            capitalize = false;
        }

        if (!Enum.TryParse<T>(buffer[..length], ignoreCase: false, out var parsed))
        {
            throw new InvalidOperationException(
                $"Unknown {typeof(T).Name} variant '{value}'.");
        }
        return parsed;
    }

    public static string ToKebab<T>(T value) where T : struct, Enum
    {
        var name = value.ToString();
        var sb = new System.Text.StringBuilder(name.Length + 4);
        for (var i = 0; i < name.Length; i++)
        {
            var ch = name[i];
            if (i > 0 && char.IsUpper(ch))
            {
                sb.Append('-');
            }
            sb.Append(char.ToLowerInvariant(ch));
        }
        return sb.ToString();
    }
}

File: examples/companion-packages/php-composer-app/src/Variants.php
<?php

declare(strict_types=1);

namespace AgentXM\Examples\PawMatch;

use InvalidArgumentException;

final class Variants
{
    public const PET_CARD_STYLES = ['compact', 'detailed', 'playful'];
    public const MATCH_STRATEGIES = ['popularity', 'match-quiz', 'longest-stay'];
    public const MATCH_DEPTHS = ['short', 'standard', 'thorough'];
    public const DONATE_FOCUSES = ['all', 'shelters', 'rescue'];

    public static function parsePetCardStyle(string $value): string
    {
        return self::requireOneOf($value, self::PET_CARD_STYLES, 'PetCardStyle');
    }

    public static function parseMatchStrategy(string $value): string
    {
        return self::requireOneOf($value, self::MATCH_STRATEGIES, 'MatchStrategy');
    }

    public static function parseMatchDepth(string $value): string
    {
        return self::requireOneOf($value, self::MATCH_DEPTHS, 'MatchDepth');
    }

    public static function parseDonateFocus(string $value): string
    {
        return self::requireOneOf($value, self::DONATE_FOCUSES, 'DonateFocus');
    }

    /**
     * @param list<string> $allowed
     */
    private static function requireOneOf(string $value, array $allowed, string $label): string
    {
        if (! in_array($value, $allowed, true)) {
            throw new InvalidArgumentException("Unknown {$label} variant '{$value}'.");
        }

        return $value;
    }
}
